code/src/context/i18n/I18nProvider.test.tsx

```diff
@@ -0,0 +1,125 @@
+import { fireEvent, render, renderHook, screen } from "@testing-library/react";
+
+import { I18nProvider, useLocaleContext } from "@/context/i18n/I18nProvider";
+import renderWithProviders from "@/utils/render-with-providers/renderWithProviders";
+
+const LocaleDisplay = () => {
+  const { locale } = useLocaleContext();
+  return <div data-testid="localeDisplay">{locale}</div>;
+};
+
+const LocaleChanger = () => {
+  const { setLocale } = useLocaleContext();
+  return (
+    <>
+      <button data-testid="setLocaleEn" onClick={() => setLocale("en")}>
+        Set Locale to EN
+      </button>
+      <button data-testid="setLocaleUk" onClick={() => setLocale("uk")}>
+        Set Locale to UK
+      </button>
+    </>
+  );
+};
+const mockAndRender = (locale: string | null) => {
+  Storage.prototype.getItem = jest.fn(() => locale);
+
+  render(<I18nPlayground />);
+};
+
+const I18nPlayground = () => {
+  return (
+    <I18nProvider>
+      <LocaleDisplay />
+      <LocaleChanger />
+    </I18nProvider>
+  );
+};
+
+describe("I18nProvider", () => {
+  let localeDisplay: HTMLElement;
+  let setUkButton: HTMLElement;
+  let setEnButton: HTMLElement;
+
+  describe("Locale display and changer functionality", () => {
+    beforeEach(() => {
+      renderWithProviders(<I18nPlayground />);
+
+      localeDisplay = screen.getByTestId("localeDisplay");
+      setUkButton = screen.getByTestId("setLocaleUk");
+      setEnButton = screen.getByTestId("setLocaleEn");
+    });
+
+    test("initially sets the locale to 'en'", () => {
+      expect(localeDisplay).toHaveTextContent("en");
+    });
+
+    test("changes locale to 'uk' when setLocaleUk is clicked", () => {
+      fireEvent.click(setUkButton);
+
+      expect(localeDisplay).toHaveTextContent("uk");
+    });
+
+    test("changes locale back to 'en' when setLocaleEn is clicked after setting it to 'uk'", () => {
+      fireEvent.click(setUkButton);
+      fireEvent.click(setEnButton);
+
+      expect(localeDisplay).toHaveTextContent("en");
+    });
+  });
+
+  describe("I18nProvider and getInitialLocale", () => {
+    afterEach(() => {
+      jest.restoreAllMocks();
+    });
+
+    test("sets the locale correctly based on browser locale", () => {
+      Object.defineProperty(navigator, "language", {
+        value: "uk-UA",
+        configurable: true
+      });
+
+      renderWithProviders(<I18nPlayground />);
+
+      const localeDisplay = screen.getByTestId("localeDisplay");
```

Comment:
what do you think if we use `getByTestId` on the `beforeEach` test can we remove this ?
And also look at the other tests where it's present. I guess it can be removed

Reply:
We cannot move it to before each because these test cases differ. As you can see for some tests I must mock and render, for some I don't need this option. Also, for some cases, I need to define the properties of the navigator object, so I don't know how to use beforeEach here, they are different

![image](https://github.com/user-attachments/assets/e23ecc3a-bc61-4916-b52e-11367cb1c4ad)

Reply:
> We cannot move it to before each because these test cases differ. As you can see for some tests I must mock and render, for some I don't need this option. Also, for some cases, I need to define the properties of the navigator object, so I don't know how to use beforeEach here, they are different
> 
> ![image](https://private-user-images.githubusercontent.com/118522791/350510360-e23ecc3a-bc61-4916-b52e-11367cb1c4ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0MTIxNzQsIm5iZiI6MTcyMTQxMTg3NCwicGF0aCI6Ii8xMTg1MjI3OTEvMzUwNTEwMzYwLWUyM2VjYzNhLWJjNjEtNDkxNi1iNTJlLTExMzY3Y2IxYzRhZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQxNzU3NTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Zjg2ODU5NjExY2U1MThhZGZlZGVhMDMxM2EwNWNiZWIxZTQ0N2IyMmQwOGNkNzM5ZjU3MzQ3ZWZkZTY0MjJkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Muy0PBNTaaN8nzCf_aChn1cRCFXpxiT0nSfOogpR8ag)

Look. I'm not talk about moving it into `beforeEach` etc.
You are already has this on the `beforeEach`
<img width="482" alt="image" src="https://github.com/user-attachments/assets/be51c432-abeb-4872-a855-3f846d84d8f8">

Can remove that variable. I can show other your cases without using it just expect
<img width="393" alt="image" src="https://github.com/user-attachments/assets/b8010006-1754-4837-9c0f-57305d07b77e">

I mean that can remove the variable here